### PR TITLE
uc-crux-llvm: Some support for arrays of unbounded size

### PR DESCRIPTION
**Pull request by langston-barrett**
The included test program causes a panic before this commit, now it mostly works.

### CODE REVIEW

File: uc-crux-llvm/src/UCCrux/LLVM/Shape.hs [hunk 1]
```diff
@@ -266,6 +275,7 @@ ppShapeSeekError =
 
 -- | Modify the 'Shape' at a given 'Cursor'
 modifyA ::
+  forall f m a tag inTy atTy.
```
Comment (travitch) [inline uc-crux-llvm/src/UCCrux/LLVM/Shape.hs:278]: Does anything introduced by this change actually require the type variables to be brought into scope?

File: uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs [hunk 2]
```diff
@@ -46,6 +46,7 @@ toMemType =
     FTNonVoidFuncPtrRepr retRepr argsRepr -> funType (Just retRepr) argsRepr
     FTOpaquePtrRepr _ident -> PtrType OpaqueType
     FTArrayRepr natRepr fullTypeRepr -> ArrayType (natValue natRepr) (toMemType fullTypeRepr)
+    FTUnboundedArrayRepr fullTypeRepr -> ArrayType 0 (toMemType fullTypeRepr)
```
Comment (travitch) [inline uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs:49]: Is it always going to be a zero here? Is this just the only representation in Crucible-llvm memory types?

**Review by travitch — APPROVED** [review]
This looks fine to me. It would be helpful to note in the merge comment that:
1. The size of arrays in type specs is now `Maybe Nat`, with `Nothing` indicating unbounded
2. I believe only zero-sized unbounded arrays are currently supported (the rest is future work)?

File: uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs (hunk 2 above)
Comment (langston-barrett) [reply uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs:49]: I'm not totally sure I understand the second part of your question, but yeah, this is how `MemType` represents these unbounded/unsized array types.

File: uc-crux-llvm/src/UCCrux/LLVM/Shape.hs (hunk 1 above)
Comment (langston-barrett) [reply uc-crux-llvm/src/UCCrux/LLVM/Shape.hs:278]: Whoops, that's leftover from an earlier draft! Will remove.

**Comment by langston-barrett** [comment]
Hmm, the CI failures seem unrelated:
```
+ [[ -x bin/yices ]]
+ command -v yices
Error: Process completed with exit code 1.
```

**Comment by RyanGlScott** [comment]
I've experienced the Ubuntu failure as well in a `saw-script` PR. It appears that the site which hosts Yices bindists (https://yices.csl.sri.com/) is down, and trying to run something like `curl -o yices.tar.gz -sl "https://yices.csl.sri.com/releases/2.6.2/yices-2.6.2-x86_64-pc-linux-gnu-static-gmp.tar.gz"` will simply error out.

**Pull request merged**